File: examples/buttons/server/django/script/check_shared_boundary.py
#!/usr/bin/env python
"""THE SHARED BOUNDARY, enforced (the Rails demo's check-shared-boundary.rb twin).

examples/buttons/shared/ is named so that a wrong import is visible in the diff:
  client/          React + Mantine + mobx-keystone — this stack's UI.
  client-vanilla/  the no-framework host ONLY.
  server-node/     SQLite and Express — the Node stacks ONLY.
Django has the ORM and its own views, so its client may import
shared/shop-types.ts, shared/http.ts, shared/bootstrap.ts,
shared/checkout-resume.ts, shared/shop.css and shared/client/**, and never
shared/server-node/** or shared/client-vanilla/**.
"""

from __future__ import annotations

import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SOURCES = [*ROOT.glob("src/**/*.ts"), *ROOT.glob("src/**/*.tsx"), ROOT / "vite.config.ts"]
FORBIDDEN = ("server-node", "client-vanilla")

violations: list[str] = []
for path in SOURCES:
    for number, line in enumerate(path.read_text(encoding="utf-8").splitlines(), start=1):
        if "shared/" not in line:
            continue
        for directory in FORBIDDEN:
            if f"shared/{directory}/" in line:
                violations.append(f"{path.relative_to(ROOT)}:{number} imports shared/{directory}/")

if violations:
    print(
        "The Django demo may only import shared/shop-types.ts, shared/http.ts and shared/client/**:"
    )
    for violation in violations:
        print(f"- {violation}")
    sys.exit(1)

print(
    f"Shared-boundary check passed: {len(SOURCES)} client files, no shared/server-node or "
    "shared/client-vanilla imports."
)

File: examples/buttons/server/laravel/script/check-shared-boundary.php
<?php

declare(strict_types=1);

// THE SHARED BOUNDARY, enforced. examples/buttons/shared/ is laid out so a
// wrong import is visible in the diff rather than discovered at build time:
//
//   client/          React + Mantine + mobx-keystone. Rails, Next.js, Laravel.
//   client-vanilla/  the no-framework host ONLY.
//   server-node/     SQLite and Express. The Node stacks ONLY.
//
// This app has Eloquent and its own controllers, so its browser code may
// import shared/shop-types.ts, shared/http.ts, shared/bootstrap.ts,
// shared/checkout-resume.ts, shared/shop.css and shared/client/** — and never
// shared/server-node/** or shared/client-vanilla/**.

$root = dirname(__DIR__);
$files = [];
foreach (['resources/js', 'vite'] as $dir) {
    $iterator = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($root.'/'.$dir, FilesystemIterator::SKIP_DOTS));
    foreach ($iterator as $file) {
        if (preg_match('/\.(ts|tsx|js|jsx)$/', $file->getFilename()) === 1) {
            $files[] = $file->getPathname();
        }
    }
}
$files[] = $root.'/vite.config.ts';

$violations = [];
foreach ($files as $path) {
    foreach (file($path, FILE_IGNORE_NEW_LINES) ?: [] as $index => $line) {
        foreach (['server-node', 'client-vanilla'] as $forbidden) {
            if (str_contains($line, "shared/{$forbidden}/")) {
                $violations[] = str_replace($root.'/', '', $path).':'.($index + 1)." imports shared/{$forbidden}/";
            }
        }
    }
}

if ($violations !== []) {
    fwrite(STDERR, "The Laravel demo may only import shared/shop-types.ts, shared/*.ts and shared/client/**:\n");
    foreach ($violations as $violation) {
        fwrite(STDERR, "- {$violation}\n");
    }
    exit(1);
}

echo 'Shared-boundary check passed: '.count($files)." client files, no shared/server-node or shared/client-vanilla imports.\n";
